File: ts/test/c2021_01/c2021_01_w1.ts
import {expect} from 'chai';

// noinspection JSUnusedLocalSymbols
/**
 * ### Palindrome Permutation
 * {@link https://leetcode.com/explore/featured/card/january-leetcoding-challenge-2021/579/week-1-january-1st-january-7th/3588/}
 */
function canPermutePalindrome(s: string): boolean {
    const map = new Map<string, boolean>();
    let odds = 0;
    for (let i = 0; i < s.length; i++) {
        const prev = map.get(s[i]);
        if (prev) {
            map.set(s[i], false);
            odds--;
        } else {
            map.set(s[i], true);
            odds++;
        }
    }
    return odds < 2;
}

// noinspection JSUnusedLocalSymbols
function canPermutePalindromeSet(s: string): boolean {
    const set = new Set<string>();
    for (let i = 0; i < s.length; i++) {
        if (set.has(s[i])) {
            set.delete(s[i]);
        } else {
            set.add(s[i]);
        }
    }
    return set.size < 2;
}

describe('c2021_01_w1 Palindrome Permutation', () => {
    it('Example 1: ("code") -> false', () => {
        expect(canPermutePalindrome('code')).to.be.false;
    });
    it('Example 2: ("aab") -> true', () => {
        expect(canPermutePalindrome('aab')).to.be.true;
    });
    it('Example 3: ("carerac") -> true', () => {
        expect(canPermutePalindrome('carerac')).to.be.true;
    });

    it('("a".repeat(500000)) -> true', () => {
        expect(canPermutePalindrome('a'.repeat(500000))).to.be.true;
    });
});

File: rust/2020_2021-09/src/c2021/c2021_01/c2021_01_w1.rs
#![allow(dead_code)]

/// ### Palindrome Permutation
/// https://leetcode.com/explore/featured/card/january-leetcoding-challenge-2021/579/week-1-january-1st-january-7th/3588/
struct Solution;

//noinspection DuplicatedCode
impl Solution {
    pub fn can_permute_palindrome(s: String) -> bool {
        use std::collections::HashMap;

        s.chars()
            .fold((0, HashMap::new()), |(odds, mut acc), c| {
                let e = acc.entry(c).or_insert(false);
                if *e {
                    *e = false;
                    (odds - 1, acc)
                } else {
                    *e = true;
                    (odds + 1, acc)
                }
            }).0 < 2
    }

    pub fn can_permute_palindrome_hash_map_char_is_odd(s: String) -> bool {
        use std::collections::HashMap;

        s.chars()
            .fold(HashMap::new(), |mut acc, c| {
                let e = acc.entry(c).or_insert(false);
                *e = if *e { false } else { true };
                acc
            })
            .values()
            .fold(0, |acc, &v| if v { acc + 1 } else { acc }) < 2
    }

    pub fn can_permute_palindrome_hash_map_char_count(s: String) -> bool {
        use std::collections::HashMap;

        s.chars()
            .fold(HashMap::new(), |mut counts, c| {
                *counts.entry(c).or_insert(0) += 1;
                counts
            })
            .values()
            .fold(0, |acc, &v| if v % 2 == 0 { acc } else { acc + 1 }) < 2
    }

    pub fn can_permute_palindrome_hash_set(s: String) -> bool {
        use std::collections::HashSet;

        s.chars()
            .fold(HashSet::new(), |mut acc, c| {
                if !acc.insert(c) { acc.remove(&c); }
                acc
            }).len() < 2
    }

    pub fn can_permute_palindrome_hash_set_contains_remove_insert(s: String) -> bool {
        use std::collections::HashSet;

        s.chars()
            .fold(HashSet::new(), |mut acc, c| {
                if acc.contains(&c) {
                    acc.remove(&c);
                } else {
                    acc.insert(c);
                }
                acc
            }).len() < 2
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example1_s_code_is_false() {
        let s = "code".to_string();
        assert!(!Solution::can_permute_palindrome(s));
    }

    #[test]
    fn example2_s_aab_is_true() {
        let s = "aab".to_string();
        assert!(Solution::can_permute_palindrome(s));
    }

    #[test]
    fn example3_s_carerac_is_true() {
        let s = "carerac".to_string();
        assert!(Solution::can_permute_palindrome(s));
    }

    #[test]
    fn s_a_repeat50000_is_true() {
        let s = "a".repeat(50000);
        assert!(Solution::can_permute_palindrome(s));
    }

    #[test]
    fn s_a_to_z_repeat1924_50024_is_true() {
        let s = ('a'..='z').collect::<String>().repeat(1924);
        assert!(Solution::can_permute_palindrome(s));
    }

    #[test]
    fn s_a_to_z_repeat1925_50050_is_false() {
        let s = ('a'..='z').collect::<String>().repeat(1925);
        assert!(!Solution::can_permute_palindrome(s));
    }
}
